spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala

```diff
@@ -2348,6 +2348,27 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
               .build()
           }
 
+        // datafusion's make_array only supports nullable element types
+        case array @ CreateArray(children, _) if array.dataType.containsNull =>
+          val childExprs = children.map(exprToProto(_, inputs, binding))
+          val dataType = serializeDataType(array.dataType)
```

Comment:
I could "fake" the datatype here when the element isn't nullable and tell DataFusion it is to get it to work, but I wasn't sure if that would have unintended downstream consequences. I can try to update DataFusion at some point to support non-nullable elements if all children are non-nullable.

Reply:
> I can try to update DataFusion at some point to support non-nullable elements if all children are non-nullable.

Does there exist a datafusion issue for this? Otherwise I think it would be good to create one to track the issue.

Reply:
Probably should be but was waiting to see if this was even the right way to use a ScalarUDF. Looking at it some more I'm not sure how it could even be updated with the way it currently works, since `ScalarUDFImpl.return_type` just has `DataType` and not `Field` to know whether the elements are nullable or not. Still learning how DataFusion works. Should I somehow be using the other thing created by `make_udf_expr_and_func`:
```
datafusion_functions_nested::make_array
pub fn make_array(arg: Vec<datafusion_expr::Expr>) -> datafusion_expr::Expr
```
?

Reply:
I am also still learning more on Datafusion. But my understanding aligns with this comment you made:

> I'm not sure how it could even be updated with the way it currently works, since ScalarUDFImpl.return_type just has DataType and not Field to know whether the elements are nullable or not.

My understanding is that an API breakage would be needed in Datafusion to make it possible to implement make_array with correct nullability for the element.

Reply:
Hmm looking a little more there's a chance I might be wrong. I think `return_type` is only used to infer the return type if it's not specified? I think the real issue is that `invoke` (which uses `fn array_array`) under the hood, returns an `Array` with an always nullable element. But here we have the `ArrayRef`'s which have an `is_nullable` attribute, so it might be possible. I'll try to test that out to see if that's the case. The actual error you get is:
```
Cause: org.apache.comet.CometNativeException: Invalid argument error: column types must match schema types, expected List(Field { name: "item", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }) but found List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }) at column index 0
```

Reply:
More digging made me realize this is a somewhat larger issue with ScalarUDFs, in that they don't support setting nullability at all (every ScalarUDF is assumed to be nullable). So how has this been handled elsewhere if at all? Is the best approach just to "pretend" the column is nullable for DataFusion, knowing that logically it should not contain any nulls and keep it as non-nullable on the Spark side? Otherwise any expression backed by a ScalarUDF can't support non-nullable expressions.